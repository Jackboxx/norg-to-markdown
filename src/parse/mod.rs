use self::{parse_headings::parse_headings, parse_ul::parse_ul};

pub mod parse_headings;
pub mod parse_highlighting;
pub mod parse_ul;

pub fn parse(line: &str) -> Result<String, regex::Error> {
    let headings = parse_headings(line)?;
    parse_ul(&headings)
}

#[cfg(test)]
mod tests {
    use crate::parse::parse;

    #[test]
    fn test_parse() {
        let text = "
* heading
*fake heading
- bullet point 1
- bullet point 2
- bullet point 3
-- sub bullet point 1
--- sub sub bullet point 1
--- sub sub bullet point 2
---- sub sub bullet point 2
-fake bullet point

** to do 
nothing";

        let res: String = text
            .lines()
            .map(|l| parse(l).unwrap_or("???".to_owned()))
            .collect::<Vec<String>>()
            .join("\n");

        insta::assert_snapshot!(res);
    }
}
