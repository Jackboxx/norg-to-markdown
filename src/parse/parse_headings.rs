use regex::Regex;

pub fn replace_headings(line: &str) -> Result<String, regex::Error> {
    let regex = Regex::new(r"(?m:^\*{1,5}\s)")?;
    let amount = match regex.captures(line) {
        Some(cap) => match cap.get(0) {
            Some(mat) => mat.as_str().trim_end().len(),
            None => 0,
        },
        None => 0,
    };

    let arr = vec!["#"; amount];
    Ok(regex.replace(line, arr.join("") + " ").to_string())
}

#[cfg(test)]
mod tests {
    use super::replace_headings;

    #[test]
    fn test_replace_headings() {
        assert_eq!(replace_headings("text"), Ok("text".to_owned()));
        assert_eq!(replace_headings("* heading1"), Ok("# heading1".to_owned()));
        assert_eq!(replace_headings("*heading2"), Ok("*heading2".to_owned()));
        assert_eq!(
            replace_headings("** heading large"),
            Ok("## heading large".to_owned())
        );
        assert_eq!(
            replace_headings("not a * heading"),
            Ok("not a * heading".to_owned())
        );

        let text = "
* heading
    - bullet point 1
    - bullet point 2
    - bullet point 3
** to do 
nothing";

        let res: String = text
            .lines()
            .map(|l| replace_headings(l).unwrap_or("???".to_owned()))
            .collect::<Vec<String>>()
            .join("\n");

        insta::assert_snapshot!(res);
    }
}
