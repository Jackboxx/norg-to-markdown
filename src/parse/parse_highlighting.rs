use regex::Regex;

pub fn parse_highlight(line: &str) -> Result<String, regex::Error> {
    let regex = Regex::new(r"")?;

    todo!()
}

#[cfg(test)]
mod tests {
    use crate::parse::parse_highlighting::parse_highlight;

    #[test]
    fn test_parse_highlight() {
        // bold
        assert_eq!(parse_highlight("text"), Ok("text".to_owned()));
        assert_eq!(parse_highlight("*text*"), Ok("**text**".to_owned()));
        assert_eq!(
            parse_highlight("some text with *bold*!"),
            Ok("some text with **bold**!".to_owned())
        );
        assert_eq!(
            parse_highlight("some *text with bold*!"),
            Ok("some **text with bold**!".to_owned())
        );
        assert_eq!(
            parse_highlight("some * text with no bold*!"),
            Ok("some * text with bold*!".to_owned())
        );

        // italic
        assert_eq!(parse_highlight("text"), Ok("text".to_owned()));
        assert_eq!(parse_highlight("/text/"), Ok("*text*".to_owned()));
        assert_eq!(
            parse_highlight("some text with /bold/!"),
            Ok("some text with *bold*!".to_owned())
        );
        assert_eq!(
            parse_highlight("some /text with bold/!"),
            Ok("some *text with bold*!".to_owned())
        );
        assert_eq!(
            parse_highlight("some / text with no bold/!"),
            Ok("some / text with bold/!".to_owned())
        );

        // bold italic
        assert_eq!(parse_highlight("text"), Ok("text".to_owned()));
        assert_eq!(parse_highlight("/*text*/"), Ok("***text***".to_owned()));
        assert_eq!(parse_highlight("*/text/*"), Ok("***text***".to_owned()));
        assert_eq!(parse_highlight("*/text*/"), Ok("*/text*/".to_owned()));
        assert_eq!(
            parse_highlight("some text with /bold/!"),
            Ok("some text with *bold*!".to_owned())
        );
    }
}
