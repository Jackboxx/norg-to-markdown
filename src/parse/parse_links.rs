use regex::Regex;

pub fn parse_web_links(line: &str) -> Result<String, regex::Error> {
    let regex = Regex::new(r"\{(.+):\}")?;
    Ok(regex.replace_all(line, "[$1]($1)").to_string())
}
pub fn parse_file_links(line: &str) -> Result<String, regex::Error> {
    let regex = Regex::new(r"\{:(.+):\}")?;
    Ok(regex.replace_all(line, "[$1]($1.md)").to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_web_links() {
        assert_eq!(parse_web_links("text"), Ok("text".to_owned()));
        assert_eq!(parse_web_links("- list item"), Ok("- list item".to_owned()));
        assert_eq!(parse_web_links("{link:}"), Ok("[link](link)".to_owned()));
        assert_eq!(parse_web_links(":{link}:"), Ok(":{link}:".to_owned()));
        assert_eq!(parse_web_links("{:link}:"), Ok("{:link}:".to_owned()));
    }

    #[test]
    fn test_parse_file_links() {
        assert_eq!(parse_file_links("text"), Ok("text".to_owned()));
        assert_eq!(
            parse_file_links("- list item"),
            Ok("- list item".to_owned())
        );
        assert_eq!(
            parse_file_links("{:link:}"),
            Ok("[link](link.md)".to_owned())
        );
        assert_eq!(parse_file_links(":{link}:"), Ok(":{link}:".to_owned()));
        assert_eq!(parse_file_links("{:link}:"), Ok("{:link}:".to_owned()));
    }
}
