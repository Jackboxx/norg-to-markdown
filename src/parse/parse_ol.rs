use regex::Regex;

use crate::regex::captures::lenght_of_nth_capture;

pub fn parse_ol(line: &str) -> Result<String, regex::Error> {
    let regex = Regex::new(r"^(\s*)(~{1,7})\s")?;
    let amount = lenght_of_nth_capture(&regex, line, 1) / 3;

    let placeholder = vec!["#"; amount + 1];
    let tabs = vec!["\t"; amount];

    Ok(regex
        .replace(
            line,
            &format!("{}[{}]. ", tabs.join(""), placeholder.join("")),
        )
        .to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_ul() {
        assert_eq!(parse_ol("text"), Ok("text".to_owned()));
        assert_eq!(parse_ol("~item"), Ok("~item".to_owned()));
        assert_eq!(parse_ol("~ item"), Ok(format!("[#]. item")));
        assert_eq!(parse_ol("~~ item"), Ok(format!("[#]. item")));
        assert_eq!(parse_ol("   ~~ item"), Ok(format!("\t[##]. item")));
        assert_eq!(parse_ol("~~item"), Ok("~~item".to_owned()));
        assert_eq!(parse_ol("   ~~~~ item"), Ok(format!("\t[##]. item")));
        assert_eq!(
            parse_ol("~~~~~~~~~~ item"),
            Ok("~~~~~~~~~~ item".to_owned())
        );
    }
}
