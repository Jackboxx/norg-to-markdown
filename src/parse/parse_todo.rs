use regex::Regex;

pub fn parse_todo(line: &str) -> Result<String, regex::Error> {
    let regex = Regex::new(r"^(\s*-{1,7}\s)\(([ x?!+=_-]?)\)(.*)")?;
    let status = match regex.captures(line) {
        Some(cap) => match cap.get(2) {
            Some(mat) => mat.as_str(),
            None => "",
        },
        None => "",
    };

    let done = if status == "x" { "x" } else { "" };
    let suffix = match status {
        " " => " (didn't do)",
        "?" => " (needs further input/clarification)",
        "!" => " (urgent)",
        "+" => " (recurring)",
        "-" => " (in-progress/pending)",
        "=" => " (on hold)",
        "_" => " (put down/cancelled)",
        _ => "",
    };

    Ok(regex
        .replace(line, format!("$1[{done}]$3{suffix}"))
        .to_string())
}

#[cfg(test)]
mod tests {
    use super::parse_todo;

    #[test]
    fn test_parse_todo() {
        assert_eq!(parse_todo("text"), Ok("text".to_owned()));
        assert_eq!(parse_todo("- item"), Ok("- item".to_owned()));
        assert_eq!(
            parse_todo("- () todo item"),
            Ok("- [] todo item".to_owned())
        );
        assert_eq!(
            parse_todo("- (x) todo item"),
            Ok("- [x] todo item".to_owned())
        );
        assert_eq!(
            parse_todo("- (_) todo item"),
            Ok("- [] todo item (put down/cancelled)".to_owned())
        );
        assert_eq!(
            parse_todo("- (=) todo item"),
            Ok("- [] todo item (on hold)".to_owned())
        );
        assert_eq!(
            parse_todo("- (r) todo item"),
            Ok("- (r) todo item".to_owned())
        );
    }
}
