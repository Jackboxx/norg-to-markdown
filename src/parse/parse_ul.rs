use regex::Regex;

use crate::regex::captures::lenght_of_nth_capture;

pub fn parse_ul(line: &str) -> Result<String, regex::Error> {
    let regex = Regex::new(r"^-{1,7}\s")?;
    let amount = lenght_of_nth_capture(&regex, line, 0);

    let arr = vec!["\t"; (amount as i64 - 1).clamp(0, i64::MAX) as usize];
    Ok(regex
        .replace(line, &format!("{}- ", arr.join("")))
        .to_string())
}

#[cfg(test)]
mod tests {
    use crate::parse::parse_ul::parse_ul;

    #[test]
    fn test_parse_ul() {
        assert_eq!(parse_ul("text"), Ok("text".to_owned()));
        assert_eq!(parse_ul("-item"), Ok("-item".to_owned()));
        assert_eq!(parse_ul("- item"), Ok("- item".to_owned()));
        assert_eq!(parse_ul("-- item"), Ok("\t- item".to_owned()));
        assert_eq!(parse_ul("--- item"), Ok("\t\t- item".to_owned()));
        assert_eq!(parse_ul("--item"), Ok("--item".to_owned()));
        assert_eq!(parse_ul("---- item"), Ok("\t\t\t- item".to_owned()));
        assert_eq!(
            parse_ul("---------- item"),
            Ok("---------- item".to_owned())
        );
    }
}
