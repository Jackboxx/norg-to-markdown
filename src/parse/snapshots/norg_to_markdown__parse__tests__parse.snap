---
source: src/parse/mod.rs
expression: res
---

# heading
*fake heading
- bullet point 1
- bullet point 2
- bullet point 3
	- sub bullet point 1
		- sub sub bullet point 1
		- sub sub bullet point 2
			- sub sub bullet point 2
-fake bullet point

## to do 
nothing
